common/beerocks/bwl/shared/nl80211_client_factory.cpp

```diff
@@ -0,0 +1,34 @@
+/* SPDX-License-Identifier: BSD-2-Clause-Patent
+ *
+ * Copyright (c) 2020 MaxLinear
+ *
+ * This code is subject to the terms of the BSD+Patent license.
+ * See LICENSE file for more details.
+ */
+
+#include <bwl/nl80211_client_factory.h>
+
+#include "nl80211_client_impl.h"
+
+#include <bcl/beerocks_backport.h>
+
+#include <easylogging++.h>
+
+namespace bwl {
+
+std::unique_ptr<nl80211_client> nl80211_client_factory::create_instance()
+{
+    // Create NL80211 socket to communicate with WiFi driver
+    std::shared_ptr<nl80211_socket> socket = std::make_shared<bwl::nl80211_socket>();
+
+    // Connect NL80211 socket (it will be automatically closed when it is not needed any more)
+    if (!socket->connect()) {
+        LOG(ERROR) << "Failed to connect nl80211 socket!";
+        return nullptr;
+    }
+
+    // Create NL80211 client to send standard NL80211 commands to WiFi driver using NL80211 socket
+    return std::make_unique<nl80211_client_impl>(socket);
+}
```

Comment:
@rmelotte your wish is my command ;-)

https://www.strawpoll.me/19474302

Reply:
I think google can provide plenty of reasons why using exceptions in embedded is bad, and plenty of why exceptions are great.
If using exceptions makes our binaries larger - then its off the table.
But even if its not, the fact remains that prplmesh is built with a strict "no exceptions" - so adding them now will just make the code a mess.
Bottom line, my vote is not to add exceptions but I'm happy to hear other suggestions.

![3r5l41](https://user-images.githubusercontent.com/13386849/75697882-6403df00-5cb6-11ea-8533-4b3c64e59230.jpg)

Reply:
AFAIK exception support in C++ causes overhead even if they are not used. 
So we should either use them everywhere or not use them anywhere (if performance is important to us).

Reply:
I don't want to start a war flame here about the use exceptions but, just for the record, I like them and therefore voted "yes".
As an advantage and just because @rmelotte has mentioned it right here, we shouldn't bother about if an error message has been printed to the log or not: log it when you catch it, no need to clutter the code with log calls everywhere. 
I also understand @tomereli's position: use of exceptions is something to be decided at the beginning of the project and if we added them now, we'd end up having prplMess instead of prplMesh.

Reply:
Regarding the use of a factory: its just a static method to create a client with which to send commands to WiFi driver. There are two different factories: one for the Dummy flavor of the BWL which creates dummy clients and one for the rest of flavors which creates real clients (those that use a socket to communicate with the driver).
Socket is connected at the beginning and stays open all the time for performance reasons. Documentation of `nl80211_client_impl` warns that socket must be connected before calling its methods, but it could have been implemented in such a way that socket was connected and closed on each call (or even created/connected/used/closed/destroyed on each call).